### PR TITLE
reauthorize when tokens expire

### PR DESCRIPTION
**Pull request by erksmit**
fix an issue in waifu command where waifulistapi tokens would expire, and never get refreshed

### CODE REVIEW

**Review by JustAnyones — CHANGES_REQUESTED** [review]
I believe effort should be taken to prevent it from continuously failing in an infinite loop.

File: bot/cogs/models/waifulistapi.py [hunk 1]
```diff
@@ -249,15 +249,18 @@ async def get(self, endpoint: str) -> Response:
...
+    async def post(self, endpoint: str, json: dict, attempts=0) -> Response:
         """Sends a POST request to Mywaifulist API endpoint."""
         r = await self.client.post(
             API_URL + endpoint, json=json,
             headers=await self.forged_headers, cookies=await self.forged_cookies
         )
+        r.status_code = 419
         if r.status_code == 419:
+            if attempts > 2:
+                raise Exception('Re-authorization attempts failed too many times')
```
Comment (JustAnyones) [inline bot/cogs/models/waifulistapi.py:261]: It should not raise such a generic Exception, you could use APIException as defined in utils/exceptions.py

File: bot/cogs/models/waifulistapi.py [hunk 2]
```diff
@@ -249,15 +249,18 @@ async def get(self, endpoint: str) -> Response:
...
         return r
 
-    async def post(self, endpoint: str, json: dict) -> Response:
+    async def post(self, endpoint: str, json: dict, attempts=0) -> Response:
         """Sends a POST request to Mywaifulist API endpoint."""
         r = await self.client.post(
             API_URL + endpoint, json=json,
             headers=await self.forged_headers, cookies=await self.forged_cookies
         )
+        r.status_code = 419
```
Comment (JustAnyones) [inline bot/cogs/models/waifulistapi.py:258]: Code will always fail

File: bot/cogs/models/waifulistapi.py [hunk 3]
```diff
@@ -249,15 +249,18 @@ async def get(self, endpoint: str) -> Response:
         )
         return r
 
-    async def post(self, endpoint: str, json: dict) -> Response:
+    async def post(self, endpoint: str, json: dict, attempts=0) -> Response:
```
Comment (JustAnyones) [inline bot/cogs/models/waifulistapi.py:252]: Add type hints

**Review by JustAnyones — CHANGES_REQUESTED** [review]

File: bot/cogs/models/waifulistapi.py (hunk 2 above)
Comment (erksmit) [reply bot/cogs/models/waifulistapi.py:258]: How did i forget to remove that

File: bot/cogs/models/waifulistapi.py (hunk 1 above)
Comment (erksmit) [reply bot/cogs/models/waifulistapi.py:261]: i thought those were reserved for theotown api errors

File: bot/cogs/ext/commands/anime.py [hunk 4]
```diff
@@ -200,7 +201,10 @@ async def waifu(self, ctx: Context, *, selection: str = None):
                 search_data = api.search_cache.get(selection)
                 if search_data is None:
                     async with ctx.typing():
-                        search_data = await api.search(selection)
+                        try:
+                            search_data = await api.search(selection)
+                        except APIException:
+                            return await ctx.reply(embed=error("Re-authorization attempts failed too many times!"))
```
Comment (JustAnyones) [inline bot/cogs/ext/commands/anime.py:207]:
```suggestion
                        except APIException as e:
                            return await ctx.reply(embed=error(str(e)))
```

Comment (erksmit) [reply bot/cogs/ext/commands/anime.py:207]: Forgot why I didn't do that

File: bot/cogs/models/waifulistapi.py [hunk 5]
```diff
@@ -249,12 +251,17 @@ async def get(self, endpoint: str) -> Response:
...
-    async def post(self, endpoint: str, json: dict) -> Response:
+    async def post(self, endpoint: str, json: dict, attempts: int = 0) -> Response:
         """Sends a POST request to Mywaifulist API endpoint."""
         r = await self.client.post(
             API_URL + endpoint, json=json,
             headers=await self.forged_headers, cookies=await self.forged_cookies
         )
+        if r.status_code == 419:
+            if attempts > 2:
+                raise APIException('Re-authorization attempts failed too many times')
```
Comment (JustAnyones) [inline bot/cogs/models/waifulistapi.py:262]: Sounds like crack English

Maybe change it to something like
> Re-authorization attempts have failed. Try again later.

**Review by JustAnyones — COMMENTED** [review]
finally, a legit english installation

**Review by JustAnyones — APPROVED** [review]

**Pull request merged**